workers/unity/Assets/Gdk/Core/GameObjectRepresentation/EntityGameObjectLifecycleSystem.cs

```diff
@@ -0,0 +1,65 @@
+using System.Collections.Generic;
+using Unity.Collections;
+using Unity.Entities;
+using UnityEngine;
+
+namespace Improbable.Gdk.Core
+{
+    public class EntityGameObjectLifecycleSystem : ComponentSystem
+    {
+        private uint currentHandle;
+        private readonly Dictionary<uint, GameObject> entityGameObjectCache = new Dictionary<uint, GameObject>();
+
+        public struct AddedEntitiesData
+        {
+            public int Length;
+            public EntityArray Entities;
+            [ReadOnly] public ComponentArray<GameObjectReference> GameObjectReferences;
+            public SubtractiveComponent<GameObjectReferenceHandle> NoGameObjectReferenceHandle;
+        }
+
+        public struct RemovedEntitiesData
+        {
+            public int Length;
+            public EntityArray Entities;
+            [ReadOnly] public ComponentDataArray<GameObjectReferenceHandle> GameObjectReferenceHandles;
+            public SubtractiveComponent<GameObjectReference> NoGameObjectReference;
+        }
+
+        [Inject] private AddedEntitiesData addedEntitiesData;
+        [Inject] private RemovedEntitiesData removedEntitiesData;
+
+        protected override void OnUpdate()
+        {
+            for (int i = 0; i < addedEntitiesData.Length; i++)
+            {
+                var handle = currentHandle++;
+                var gameObjectReferenceHandleComponent = new GameObjectReferenceHandle { GameObjectHandle = handle };
+                PostUpdateCommands.AddComponent(addedEntitiesData.Entities[i], gameObjectReferenceHandleComponent);
+                var gameObject = addedEntitiesData.GameObjectReferences[i].GameObject;
+                entityGameObjectCache[handle] = gameObject;
+            }
+
+            for (int i = 0; i < removedEntitiesData.Length; i++)
+            {
+                var handle = removedEntitiesData.GameObjectReferenceHandles[i].GameObjectHandle;
+                GameObject gameObject;
+                if (!entityGameObjectCache.TryGetValue(handle, out gameObject))
+                {
+                    Debug.LogErrorFormat(Errors.GameObjectNotFound, handle);
```

Comment:
Under what circumstances is this object not there. Presumably it could point to "null", but the dictionary should have a value for this key.

Reply:
🤔 this could happen if the component gets added to the entity by another system or whatever creates the entity... which gives me an idea... should that component type be a private nested class of this system in that case?

Reply:
The error is one of those "This should never happen" cases.